### PR TITLE
Encryption at rest support

### PR DESCRIPTION
**Pull request by ewoutp**
# What 
This PR adds support for encrypting data stored by RocksDB when written to disk.

# How 
It adds an `EncryptedEnv` override of the `Env` class with matching overrides for sequential&random access files.
The encryption itself is done through a configurable `EncryptionProvider`. This class creates is asked to create `BlockAccessCipherStream` for a file. This is where the actual encryption/decryption is being done.
Currently there is a Counter mode implementation of `BlockAccessCipherStream` with a `ROT13` block cipher (NOTE the `ROT13` is for demo purposes only!!).

The Counter operation mode uses an initial counter & random initialization vector (IV).
Both are created randomly for each file and stored in a 4K (default size) block that is prefixed to that file. The `EncryptedEnv` implementation is such that clients of the `Env` class do not see this prefix (nor data, nor in filesize).
The largest part of the prefix block is also encrypted, and there is room left for implementation specific settings/values/keys in there.

# Testing 
To test the encryption, the `DBTestBase` class has been extended to consider a new environment variable called `ENCRYPTED_ENV`. If set, the test will setup a encrypted instance of the `Env` class to use for all tests. 
Typically you would run it like this:

```
ENCRYPTED_ENV=1 make check_some
``` 

There is also an added test that checks that some data inserted into the database is or is not "visible" on disk. With `ENCRYPTED_ENV` active it must not find plain text strings, with `ENCRYPTED_ENV` unset, it must find the plain text strings.

### CODE REVIEW

File: db/db_encryption_test.cc [hunk 1]
```diff
@@ -0,0 +1,99 @@
+//  Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
+//  This source code is licensed under the BSD-style license found in the
+//  LICENSE file in the root directory of this source tree. An additional grant
+//  of patent rights can be found in the PATENTS file in the same directory.
+//  This source code is also licensed under the GPLv2 license found in the
+//  COPYING file in the root directory of this source tree.
+//
+// Copyright (c) 2011 The LevelDB Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file. See the AUTHORS file for names of contributors.
```
Comment (sagar0) [inline db/db_encryption_test.cc]: Lines 8-10 are not needed, as it is a new file being added directly into RocksDB.

File: env/env_encryption.cc [hunk 2]
```diff
@@ -0,0 +1,907 @@
+//  Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
+//  This source code is licensed under the BSD-style license found in the
+//  LICENSE file in the root directory of this source tree. An additional grant
+//  of patent rights can be found in the PATENTS file in the same directory.
+//  This source code is also licensed under the GPLv2 license found in the
+//  COPYING file in the root directory of this source tree.
+//
+// Copyright (c) 2011 The LevelDB Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file. See the AUTHORS file for names of contributors
```
Comment (sagar0) [inline env/env_encryption.cc]: Lines 8-10 are not needed, as it is a new file being added directly into RocksDB.

File: Makefile [hunk 3]
```diff
@@ -324,6 +324,7 @@ EXPOBJECTS = $(EXP_LIB_SOURCES:.cc=.o) $(LIBOBJECTS) $(TESTUTIL)
 
 TESTS = \
 	db_basic_test \
+	db_encryption_test \
```
Comment (sagar0) [inline Makefile]: I think you also need to update TARGETS file. (It is a file we use internally at Facebook to build with buck).

**Comment by sagar0** [comment]
Thanks for your contribution! 
Seems like Travis tests failed to kickoff, and there's also a merge conflict; so could you please rebase and submit again?

File: db/db_encryption_test.cc [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+    std::string scratch;
+    scratch.reserve(fileSize);
+    Slice data;
+    status = seqFile->Read(fileSize, &data, (char*)scratch.data());
+    ASSERT_OK(status);
+
+    if (data.ToString().find("foo567") != std::string::npos) {
+      hits++; 
+      std::cout << "Hit in " << filePath << "\n";
```
Comment (sagar0) [inline db/db_encryption_test.cc]: We typically don't print to stdout in tests (there are a few exceptions in our source code though). May be these are left over from your debugging?

File: env/env_encryption.cc (hunk 2 above)
Comment (ewoutp) [inline env/env_encryption.cc]: Removed them

File: Makefile (hunk 3 above)
Comment (ewoutp) [inline Makefile]: Added it to TARGETS (please check for correctness)

File: db/db_encryption_test.cc (hunk 4 above)
Comment (ewoutp) [inline db/db_encryption_test.cc]: Commented them out.

**Comment by sagar0** [comment]
Travis `lite_build`s are failing on both linux and mac. Can you please take a look? 
Also, I just restarted linux platform_dependent test, as it failed with 'No space left on device' error. (rate_limiter test in Mac platform_dependent test suite has been consistently failing for unrelated issues. So we can disregard it for now).

**Comment by ewoutp** [comment]
@sagar0 Shall I just wrap the entire file in `#ifndef ROCKSDB_LITE`?

**Comment by ewoutp** [comment]
Only java_test is failing. Have no clue how that is related.
Any ideas?

**Comment by sagar0** [comment]
Thanks for fixing the lite build. Java test is a little flaky ... so we don't need to worry about it most of the time. I'll take a look.

File: TARGETS [hunk 5]
```diff
@@ -431,6 +431,7 @@ ROCKS_TESTS = [['merger_test', 'table/merger_test.cc', 'serial'],
  ['options_util_test', 'utilities/options/options_util_test.cc', 'serial'],
  ['dynamic_bloom_test', 'util/dynamic_bloom_test.cc', 'serial'],
  ['db_basic_test', 'db/db_basic_test.cc', 'serial'],
+ ['db_encryption_test', 'db/db_encryption_test.cc', 'serial'],
```
Comment (sagar0) [inline TARGETS]: env/env_encryption.cc needs to be added near other env*.cc files.

File: src.mk [hunk 6]
```diff
@@ -57,6 +57,7 @@ LIB_SOURCES =                                                   \
   db/write_thread.cc                                            \
   env/env.cc                                                    \
   env/env_chroot.cc                                             \
+  env/env_encryption.cc                                         \
```
Comment (sagar0) [inline src.mk]: db_encryption_test.cc should also be added to src.mk.

File: env/env_encryption.cc (hunk 2 above)
Comment (sagar0) [inline env/env_encryption.cc]: I guess you forgot to remove these lines. :)

**Comment by sagar0** [comment]
Would it also be possible to contribute a small example to show how this API can be used? It can be done in a separate pull request, when time permits.

File: TARGETS (hunk 5 above)
Comment (ewoutp) [inline TARGETS]: Done

File: src.mk (hunk 6 above)
Comment (ewoutp) [inline src.mk]: Done

**Comment by ewoutp** [comment]
Result of `make format`

```

**Comment by ewoutp** [comment]
As for sample code.

Typically you would implement a class derived from `BlockCipher` (see `ROT13Cipher`) and use it as follows:

**Comment by sagar0** [comment]
I am trying to import this PR on to Facebook's internal phabricator to run some tests, but unfortunately the import is failing. I am trying to figure out a way to move forward.

**Comment by sagar0** [comment]
@ewoutp Can you please rebase your changes on master? 
We are encountering an issue when trying to import this PR onto our internal phabricator for running additional tests, and our workflow does not let us to merge the changes without that step. 
I appreciate your patience while we are working through these kinks.

**Comment by ewoutp** [comment]
@sagar0 looks like it's already merged. If anything is still needed, let me know.

**Comment by sagar0** [comment]
@ewoutp yeah, I merged it yesterday. @sdwilsh fixed an issue on our side in the import script, and helped in the process.

**Comment by vivekpm** [comment]
Is this available through RocksDBJNI?

**Pull request closed without merging**